xmtp_id/src/associations/signature.rs

```diff
@@ -60,181 +102,253 @@ impl Clone for Box<dyn Signature> {
-                bytes: self.bytes(),
+                message: self.message.clone(),
+                public_key: self.public_key.to_vec(),
+                bytes: self.to_bytes(),
             })),
         }
     }
+
+    fn verify(&self) -> Result<(), SignatureError> {
+        let signature = ethers::types::Signature::try_from(self.to_bytes().as_slice())?;
+        signature.verify(self.message.clone(), public_key_to_address(self.public_key))?;
+        Ok(())
+    }
+}
+
+/// Convert a k256 public key to an ethereum [Address]
+pub fn public_key_to_address(public_key: [u8; 64]) -> Address {
+    let hash = keccak256(&public_key[..]);
+    Address::from_slice(&hash[12..])
 }
 
-#[allow(dead_code)]
 #[derive(Clone)]
 pub struct Erc1271Signature {
-    signature_text: String,
-    signature_bytes: Vec<u8>,
+    message: String,
+    bytes: Vec<u8>,
+    chain: u64,
     contract_address: String,
     block_number: u64,
 }
 
 impl Erc1271Signature {
     pub fn new(
-        signature_text: String,
-        signature_bytes: Vec<u8>,
+        message: String,
+        bytes: Vec<u8>,
+        chain: u64,
         contract_address: String,
         block_number: u64,
-    ) -> Self {
-        Erc1271Signature {
-            signature_text,
-            signature_bytes,
+    ) -> Result<Self, SignatureError> {
+        let sig = Erc1271Signature {
+            message,
+            bytes,
+            chain,
             contract_address,
             block_number,
-        }
+        };
+        sig.verify()?;
+        Ok(sig)
     }
 }
 
 impl Signature for Erc1271Signature {
-    fn recover_signer(&self) -> Result<MemberIdentifier, SignatureError> {
-        // TODO: Verify signature first
-        Ok(self.contract_address.clone().into())
+    fn get_kind(&self) -> &SignatureKind {
+        &SignatureKind::Erc1271
     }
 
-    fn signature_kind(&self) -> SignatureKind {
-        SignatureKind::Erc1271
+    fn get_signer(&self) -> MemberIdentifier {
+        MemberIdentifier::Address(self.contract_address.clone())
     }
 
-    fn bytes(&self) -> Vec<u8> {
-        self.signature_bytes.clone()
+    fn to_bytes(&self) -> Vec<u8> {
+        self.bytes.clone()
     }
 
     fn to_proto(&self) -> SignatureProto {
         SignatureProto {
             signature: Some(SignatureKindProto::Erc1271(Erc1271SignatureProto {
-                contract_address: self.contract_address.clone(),
+                message: self.message.clone(),
+                bytes: self.to_bytes(),
+                chain: self.chain.into(),
+                contract_address: self.contract_address.to_string(),
                 block_number: self.block_number,
-                signature: self.bytes(),
             })),
         }
     }
+
+    fn verify(&self) -> Result<(), SignatureError> {
+        let verifier = ERC1271Verifier::new(CHAIN_RPC_PROVIDER);
+
+        // wrap the async function with tokio Runtime
+        let runtime = Runtime::new().unwrap();
```

Comment:
FWIW, I don't mind adjusting the trait so that this method can be async

Reply:
Ohhh, I see. We can't do that because this is all happening inside `new(...)` and that gets run from inside synchronous methods.